File: src/main/java/net/softbell/bsh/iot/service/v1/IotActionServiceV1.java
package net.softbell.bsh.iot.service.v1;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Optional;

import javax.transaction.Transactional;

import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;

import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import net.softbell.bsh.domain.EnableStatusRule;
import net.softbell.bsh.domain.GroupRole;
import net.softbell.bsh.domain.ItemCategoryRule;
import net.softbell.bsh.domain.entity.Member;
import net.softbell.bsh.domain.entity.NodeAction;
import net.softbell.bsh.domain.entity.NodeActionItem;
import net.softbell.bsh.domain.entity.NodeItem;
import net.softbell.bsh.domain.repository.NodeActionItemRepo;
import net.softbell.bsh.domain.repository.NodeActionRepo;
import net.softbell.bsh.domain.repository.NodeItemRepo;
import net.softbell.bsh.dto.request.IotActionDto;
import net.softbell.bsh.dto.request.IotActionItemDto;
import net.softbell.bsh.service.MemberService;
import net.softbell.bsh.util.BellLog;

/**
 * @Author : Bell(bell@softbell.net)
 * @Description : IoT Node Action 서비스
 */
@Slf4j
@AllArgsConstructor
@Service
public class IotActionServiceV1
{
	// Global Field
	private final MemberService memberService;
	private final IotNodeServiceV1 nodeService;
	
	private final NodeItemRepo nodeItemRepo;
	private final NodeActionRepo nodeActionRepo;
	private final NodeActionItemRepo nodeActionItemRepo;
	
	public List<NodeItem> getAvailableNodeItem(Authentication auth)
	{
		// Field
		List<NodeItem> listNodeItem;
		
		// Init
		listNodeItem = nodeService.getCategoryNodeItems(auth, GroupRole.ACTION, ItemCategoryRule.CONTROL);
		
		// Return
		return listNodeItem;
	}
	
	public List<NodeAction> getAllNodeActions(Authentication auth)
	{
		// Field
		Member member;
		List<NodeAction> listNodeAction;
		
		// Init
		member = memberService.getMember(auth.getName());
		
		// Exception
		if (memberService.isAdmin(member))
			listNodeAction = nodeActionRepo.findAll();
		else
			listNodeAction = nodeActionRepo.findByMember(member);
		
		// Return
		return listNodeAction;
	}
	
	public NodeAction getNodeAction(Authentication auth, long actionId)
	{
		// Field
		Optional<NodeAction> optNodeAction;
		NodeAction nodeAction;
		
		// Init
		optNodeAction = nodeActionRepo.findById(actionId);
		
		// Exception
		if (!optNodeAction.isPresent())
			return null;
		
		// Load
		nodeAction = optNodeAction.get();
		
		// Return
		return nodeAction;
	}
	
	@Transactional
	public boolean createAction(Authentication auth, IotActionDto iotActionDto)
	{
		// Log
		log.info(BellLog.getLogHead() + "Creating Action (" + iotActionDto.getDescription() + ")");
		
		// Field
		Member member;
		NodeAction nodeAction;
		HashMap<Long, IotActionItemDto> mapActionItem;
		List<NodeActionItem> listNodeActionItem;
		EnableStatusRule enableStatus;
		
		// Init
		member = memberService.getMember(auth.getName());
		listNodeActionItem = new ArrayList<NodeActionItem>();
		mapActionItem = iotActionDto.getMapActionItem();
		if (iotActionDto.isEnableStatus())
			enableStatus = EnableStatusRule.ENABLE;
		else
			enableStatus = EnableStatusRule.DISABLE;
		
		// Exception
		if (member == null)
			return false;
		
		// Data Process - Action Info
		nodeAction = NodeAction.builder()
								.enableStatus(enableStatus)
								.description(iotActionDto.getDescription())
								.member(member)
									.build();

		// Data Process - Action Item Info
		if (mapActionItem != null)
		{
			mapActionItem.forEach((key, value) ->
			{
				if(value.getItemId() != null && value.getItemId() != 0)
				{
					// Field
					Optional<NodeItem> optNodeItem;
					NodeActionItem nodeActionItem;
					
					// Init
					optNodeItem = nodeItemRepo.findById(value.getItemId());
					
					// Build
					if (optNodeItem.isPresent())
					{
						nodeActionItem = NodeActionItem.builder()
														.nodeItem(optNodeItem.get())
														.itemStatus(value.getItemStatus())
														.nodeAction(nodeAction)
															.build();
						
						// List Add
						listNodeActionItem.add(nodeActionItem);
					}
				}
			});
			nodeAction.setNodeActionItems(listNodeActionItem);
		}

		// DB - Save
		nodeActionRepo.save(nodeAction);
		nodeActionItemRepo.saveAll(listNodeActionItem);
		
		// Log
		log.info(BellLog.getLogHead() + "Created Action (" + nodeAction.getActionId() + ", " + iotActionDto.getDescription() + ")");
		
		// Return
		return true;
	}
	
	@Transactional
	public boolean modifyAction(Authentication auth, long actionId, IotActionDto iotActionDto)
	{
		// Field
		Optional<NodeAction> optNodeAction;
		NodeAction nodeAction;
		HashMap<Long, IotActionItemDto> mapActionItem;
		List<NodeActionItem> listNodeActionItem;
		EnableStatusRule enableStatus;
		
		// Init
		optNodeAction = nodeActionRepo.findById(actionId);
		mapActionItem = iotActionDto.getMapActionItem();
		listNodeActionItem = new ArrayList<NodeActionItem>();
		if (iotActionDto.isEnableStatus())
			enableStatus = EnableStatusRule.ENABLE;
		else
			enableStatus = EnableStatusRule.DISABLE;
		
		// Exception
		if (!optNodeAction.isPresent())
			return false;
		
		// Load
		nodeAction = optNodeAction.get();
//		listNodeActionItem = nodeAction.getNodeActionItems();
		
		// DB - Action Item Clear
		nodeActionItemRepo.deleteAll(nodeAction.getNodeActionItems());
		nodeActionItemRepo.flush();
//		nodeActionRepo.flush();
//		for (NodeActionItem value : listNodeActionItem)
//			nodeAction.removeNodeActionItem(value);
		
		// Data Process - Item Info
		nodeAction.setNodeActionItems(null);
		nodeAction.setEnableStatus(enableStatus);
		nodeAction.setDescription(iotActionDto.getDescription());
		
		// DB - Save
		nodeActionRepo.save(nodeAction);
		
		// Data Process - Action Item Info
		if (mapActionItem != null)
		{
			mapActionItem.forEach((key, value) ->
			{
				if(value.getItemId() != null && value.getItemId() != 0)
				{
					// Field
					Optional<NodeItem> optNodeItem;
					NodeActionItem nodeActionItem;
					
					// Init
					optNodeItem = nodeItemRepo.findById(value.getItemId());
					
					// Build
					if (optNodeItem.isPresent())
					{
						nodeActionItem = NodeActionItem.builder()
														.nodeItem(optNodeItem.get())
														.itemStatus(value.getItemStatus())
														.nodeAction(nodeAction)
															.build();
						
						// List Add
						listNodeActionItem.add(nodeActionItem);
					}
				}
			});
			nodeAction.setNodeActionItems(listNodeActionItem);
		}
		
		// DB - Update
//		nodeActionRepo.save(nodeAction);
		nodeActionItemRepo.saveAll(listNodeActionItem);
		
		// Return
		return true;
	}
	
	@Transactional
	public boolean deleteAction(Authentication auth, long actionId)
	{
		// Field
		Optional<NodeAction> optNodeAction;
		NodeAction nodeAction;
		
		// Init
		optNodeAction = nodeActionRepo.findById(actionId);
		
		// Exception
		if (!optNodeAction.isPresent())
			return false;
		
		// Load
		nodeAction = optNodeAction.get();
		
		// DB - Delete
		nodeActionItemRepo.deleteAll(nodeAction.getNodeActionItems());
		nodeActionRepo.delete(nodeAction);
		
		// Return
		return true;
	}
	
	public boolean execAction(long actionId, Authentication auth)
	{
		// Field
		Member member;
		Optional<NodeAction> optNodeAction;
		boolean isSuccess = true;
		
		// Init
		member = memberService.getMember(auth.getName());
		optNodeAction = nodeActionRepo.findById(actionId);
		
		// Exception
		if (!optNodeAction.isPresent())
			return false;
		
		// Process
		isSuccess = execAction(optNodeAction.get(), member);
		
		// Return
		return isSuccess;
	}
	
	public boolean execAction(NodeAction nodeAction, Member member)
	{
		// Field
		List<NodeActionItem> listNodeActionItem;
		boolean isSuccess = true;
		
		// Exception
		if (nodeAction == null)
			return false;
		
		// Permission
		if (!memberService.isAdmin(member)) // 관리자가 아닌데
			if (!nodeAction.getMember().equals(member)) // 액션 제작자가 아니면
				return false; // 수행 중단
		
		// Load
		listNodeActionItem = nodeAction.getNodeActionItems();
		
		// Process
		for (NodeActionItem actionItem : listNodeActionItem)
			if (!nodeService.setItemValue(actionItem.getNodeItem(), actionItem.getItemStatus()))
				isSuccess = false;
		
		// Return
		return isSuccess;
	}
}

File: src/main/kotlin/net/softbell/bsh/iot/service/v1/IotActionServiceV1.kt
package net.softbell.bsh.iot.service.v1

import mu.KLogging
import net.softbell.bsh.domain.EnableStatusRule
import net.softbell.bsh.domain.GroupRole
import net.softbell.bsh.domain.ItemCategoryRule
import net.softbell.bsh.domain.entity.Member
import net.softbell.bsh.domain.entity.NodeAction
import net.softbell.bsh.domain.entity.NodeActionItem
import net.softbell.bsh.domain.entity.NodeItem
import net.softbell.bsh.domain.repository.NodeActionItemRepo
import net.softbell.bsh.domain.repository.NodeActionRepo
import net.softbell.bsh.domain.repository.NodeItemRepo
import net.softbell.bsh.dto.request.IotActionDto
import net.softbell.bsh.service.MemberService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.*
import java.util.function.BiConsumer
import javax.transaction.Transactional

/**
 * @author : Bell(bell@softbell.net)
 * @description : IoT Node Action 서비스
 */
@Service
class IotActionServiceV1 {
    // Global Field
    @Autowired private lateinit var memberService: MemberService
    @Autowired private lateinit var nodeService: IotNodeServiceV1
    @Autowired private lateinit var nodeItemRepo: NodeItemRepo
    @Autowired private lateinit var nodeActionRepo: NodeActionRepo
    @Autowired private lateinit var nodeActionItemRepo: NodeActionItemRepo

    fun getAvailableNodeItem(member: Member): List<NodeItem> {
        // Return
        if (memberService.isAdmin(member))
            return nodeService.getCategoryAllNodeItems(ItemCategoryRule.CONTROL)
        else
            return nodeService.getCategoryPrivilegesNodeItems(member, GroupRole.ACTION, ItemCategoryRule.CONTROL)
    }

    fun getNodeActions(): List<NodeAction> {
        // Return
        return nodeActionRepo.findAll()
    }

    fun getPrivilegesNodeActions(member: Member): List<NodeAction> {
        if (memberService.isAdmin(member))
            return getNodeActions()

        // Return
        return nodeActionRepo.findByMember(member)
    }

    fun getNodeAction(actionId: Long): NodeAction? {
        // Init
        val optNodeAction = nodeActionRepo.findById(actionId)

        // Exception
        if (!optNodeAction.isPresent)
            return null

        // Return
        return optNodeAction.get()
    }

    fun getPrivilegesNodeAction(member: Member, actionId: Long): NodeAction? {
        // TODO 권한 있는 노드 액션만 반환 필요
        // Init
        val optNodeAction = nodeActionRepo.findById(actionId)

        // Exception
        if (!optNodeAction.isPresent)
            return null

        // Return
        return optNodeAction.get()
    }

    @Transactional
    fun createAction(member: Member, iotActionDto: IotActionDto): Boolean {
        // Log
        logger.info("Creating Action (" + iotActionDto.description + ")")

        // Init
        val listNodeActionItem: MutableList<NodeActionItem> = ArrayList()
        val mapActionItem = iotActionDto.mapActionItem
        val enableStatus = if (iotActionDto.enableStatus)
            EnableStatusRule.ENABLE
        else
            EnableStatusRule.DISABLE

        // Data Process - Action Info
        val nodeAction = NodeAction(
                enableStatus = enableStatus,
                description = iotActionDto.description,
                member = member
        )

        // Data Process - Action Item Info
        if (mapActionItem != null) {
            mapActionItem.forEach(BiConsumer { key: Long, (_, itemId, itemStatus) ->
                if (itemId != -1L) {
                    // Init
                    val optNodeItem = nodeItemRepo.findById(itemId)

                    // Build
                    if (optNodeItem.isPresent) {
                        val nodeActionItem = NodeActionItem(
                                nodeItem = optNodeItem.get(),
                                itemStatus = itemStatus,
                                nodeAction = nodeAction
                        )

                        // List Add
                        listNodeActionItem.add(nodeActionItem)
                    }
                }
            })
            nodeAction.nodeActionItems = listNodeActionItem
        }

        // DB - Save
        nodeActionRepo.save(nodeAction)
        nodeActionItemRepo.saveAll(listNodeActionItem)

        // Log
        logger.info("Created Action (" + nodeAction.actionId + ", " + iotActionDto.description + ")")

        // Return
        return true
    }

    @Transactional
    fun modifyAction(actionId: Long, iotActionDto: IotActionDto): Boolean {
        // Init
        val optNodeAction = nodeActionRepo.findById(actionId)
        val mapActionItem = iotActionDto.mapActionItem
        val listNodeActionItem: MutableList<NodeActionItem> = ArrayList()
        val enableStatus = if (iotActionDto.enableStatus)
            EnableStatusRule.ENABLE
        else
            EnableStatusRule.DISABLE

        // Exception
        if (!optNodeAction.isPresent)
            return false

        // Load
        val nodeAction = optNodeAction.get()
        //		listNodeActionItem = nodeAction.getNodeActionItems();

        // DB - Action Item Clear
        nodeActionItemRepo.deleteAll(nodeAction.nodeActionItems)
        nodeActionItemRepo.flush()
        //		nodeActionRepo.flush();
//		for (NodeActionItem value : listNodeActionItem)
//			nodeAction.removeNodeActionItem(value);

        // Data Process - Item Info
//        nodeAction.nodeActionItems = null // TODO 이거 뭐지?
        nodeAction.enableStatus = enableStatus
        nodeAction.description = iotActionDto.description

        // Data Process - Action Item Info
        if (mapActionItem != null) {
            mapActionItem.forEach(BiConsumer { key: Long?, (_, itemId, itemStatus) ->
                if (itemId != -1L) {
                    // Init
                    val optNodeItem = nodeItemRepo.findById(itemId)

                    // Build
                    if (optNodeItem.isPresent) {
                        val nodeActionItem = NodeActionItem(
                                nodeItem = optNodeItem.get(),
                                itemStatus = itemStatus,
                                nodeAction = nodeAction
                        )

                        // List Add
                        listNodeActionItem.add(nodeActionItem)
                    }
                }
            })
            nodeAction.nodeActionItems = listNodeActionItem
        }

        // DB - Update
//		nodeActionRepo.save(nodeAction);
        nodeActionItemRepo.saveAll(listNodeActionItem)

        // Return
        return true
    }

    fun modifyPrivilegesAction(member: Member, actionId: Long, iotActionDto: IotActionDto): Boolean {
        // TODO 권한 체크 필요
        return modifyAction(actionId, iotActionDto)
    }

    @Transactional
    fun deleteAction(actionId: Long): Boolean {
        // Init
        val optNodeAction = nodeActionRepo.findById(actionId)

        // Exception
        if (!optNodeAction.isPresent)
            return false

        // Load
        val nodeAction = optNodeAction.get()

        // DB - Delete
        nodeActionItemRepo.deleteAll(nodeAction.nodeActionItems)
        nodeActionRepo.delete(nodeAction)

        // Return
        return true
    }

    fun deletePrivilegesAction(member: Member, actionId: Long): Boolean {
        // TODO 권한 체크 필요
        return deleteAction(actionId)
    }

    fun execAction(nodeAction: NodeAction): Boolean {
        // Init
        var isSuccess = true

        // Load
        val listNodeActionItem = nodeAction.nodeActionItems

        // Process
        for (actionItem in listNodeActionItem)
            if (!nodeService.setItemValue(actionItem.nodeItem, actionItem.itemStatus))
                isSuccess = false

        // Return
        return isSuccess
    }

    fun execPrivilegesAction(actionId: Long, member: Member): Boolean {
        // Init
        val optNodeAction = nodeActionRepo.findById(actionId)

        // Exception
        if (!optNodeAction.isPresent)
            return false

        // Return
        return execPrivilegesAction(optNodeAction.get(), member)
    }

    fun execPrivilegesAction(nodeAction: NodeAction, member: Member): Boolean {
        // Permission
        if (!memberService.isAdmin(member)) // 관리자가 아닌데
            if (nodeAction.member != member) // 액션 제작자가 아니면
                return false // 수행 중단

        // Return
        return execAction(nodeAction)
    }

    companion object : KLogging()
}